### PR TITLE
Austin/bookmarks

### PR DESCRIPTION
**Pull request by AAAstorga**
I made some massive changes. I def want to go back and clean up some code, but I want to get this merged in to get in sync for the demo tonight. Please let me know what should change!

### CODE REVIEW

File: Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift [hunk 1]
```diff
@@ -95,15 +97,16 @@ extension ArticleSectionController: HeartPressedDelegate, BookmarkPressedDelegat
         case .categoryCell:
             return CGSize(width: width, height: 40)
         case .titleCell:
-            let height = entry.title.height(withConstrainedWidth: width, font: UIFont.boldSystemFont(ofSize: 22)) //CLUTCH Extension thank stackoverflow gods
-             print(entry.title, " ", height)
+            let height = entry.title.height(withConstrainedWidth: width - 34, font: UIFont.boldSystemFont(ofSize: 22)) //CLUTCH Extension thank stackoverflow gods
             return CGSize(width: width, height: height + 10)
+        case .authorCell:
+            let height = entry.author?.name.height(withConstrainedWidth: width, font: UIFont(name: "Georgia", size: 13)!)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift]: add this font to the `UIFont+` file later, do a guard let?

File: Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift [hunk 2]
```diff
@@ -140,6 +148,8 @@ extension ArticleSectionController: HeartPressedDelegate, BookmarkPressedDelegat
             cell.heartDelegate = self
             cell.bookmarkDelegate = self
             cell.shareDelegate = self
+            cell.setupViews(forBookmarks: false)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift]: in the future we should check if it's been bookmarked before

File: Cornell Sun/Cornell Sun/Article Section/FeedCollectionViewController.swift [hunk 3]
```diff
@@ -34,6 +37,7 @@ class FeedCollectionViewController: ViewController, UIScrollViewDelegate {
         adapter.collectionView?.backgroundColor = UIColor(white: 241.0 / 255.0, alpha: 1.0)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Article Section/FeedCollectionViewController.swift]: add this color to `UIColor+` later

File: Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift [hunk 4]
```diff
@@ -35,17 +36,12 @@ class ArticleSectionController: ListSectionController {
...
 
-    func taptic() {
-        let generator = UIImpactFeedbackGenerator(style: .light)
-        generator.prepare()
-        generator.impactOccurred()
-    }
-
     func didPressBookmark(_ cell: MenuActionCell) {
+        let didBookmark = cell.bookmarkButton.currentImage == #imageLiteral(resourceName: "bookmark") //we should save the bookmark
         let correctBookmarkImage = cell.bookmarkButton.currentImage == #imageLiteral(resourceName: "bookmarkPressed") ? #imageLiteral(resourceName: "bookmark") : #imageLiteral(resourceName: "bookmarkPressed")
```
Comment (cdsciavolino) [inline Cornell Sun/Cornell Sun/Article Section/ArticleSectionController.swift]: could this be
`let correctBookmarkImage = didBookmark ? #imageLiteral(resourceName: "bookmarkPressed") : #imageLiteral(resourceName: "bookmark")`
? (make the line a lil shorter lol)

File: Cornell Sun/Cornell Sun/Article Section/FeedCollectionViewController.swift [hunk 5]
```diff
@@ -48,59 +52,25 @@ class FeedCollectionViewController: ViewController, UIScrollViewDelegate {
...
-                                    })
-                            } catch {
-
+                        if let post = PostObject(data: postDictionary) {
+                            if self.firstPostObject == nil {
+                                self.firstPostObject = post
                             }
-                        })
+                            if savedPostIds.contains(post.id) {
+                                post.didSave = true
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Article Section/FeedCollectionViewController.swift]: oh wait disregard comment above about saving

File: Cornell Sun/Cornell Sun/Bookmark Section/BookmarkSectionController.swift [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        case .categoryCell:
+            return CGSize(width: width, height: 40)
+        case .imageAndTitleCell:
+            return CGSize(width: width, height: 124)
+        case .actionMenuCell:
+            return CGSize(width: width, height: 35)
+        }
+    }
+
+    override func cellForItem(at index: Int) -> UICollectionViewCell {
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Bookmark Section/BookmarkSectionController.swift]: the switch statement can be refactored to context.dequeue of type [celltype...] but it's good for now

File: Cornell Sun/Cornell Sun/Bookmark Section/BookmarkSectionController.swift [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            return cell
+        }
+    }
+
+    override func didUpdate(to object: Any) {
+        entry = object as? PostObject
+    }
+
+    override func didSelectItem(at index: Int) {
+        if index != 2 {
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Bookmark Section/BookmarkSectionController.swift]: make these indices enums

File: Cornell Sun/Cornell Sun/Cells/AuthorCell.swift [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            authorLabel.text = post?.author?.name
+            timeStampLabel.text = post?.datePosted.timeAgoSinceNow()
+        }
+    }
+
+    let authorLabel: UILabel = {
+        let label = UILabel()
+        label.text = ""
+        label.numberOfLines = 1
+        label.font = UIFont(name: "Georgia", size: 13)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/AuthorCell.swift]: add these fonts to `UIFont+`

File: Cornell Sun/Cornell Sun/Cells/AuthorCell.swift [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    func setupViews() {
+        self.backgroundColor = .white
+        addSubview(authorLabel)
+        addSubview(timeStampLabel)
+        authorLabel.snp.makeConstraints { (make) in
+            make.top.equalToSuperview()
+            make.left.equalToSuperview().inset(17)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/AuthorCell.swift]: make this 17 inset a constant

File: Cornell Sun/Cornell Sun/Cells/AuthorCell.swift [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        timeStampLabel.snp.makeConstraints { (make) in
+            make.top.equalToSuperview()
+            make.right.equalToSuperview().inset(17)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/AuthorCell.swift]: make this a constant too

File: Cornell Sun/Cornell Sun/Cells/BookmarkCell.swift [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        self.backgroundColor = .white
+        addSubview(imageView)
+        addSubview(titleLabel)
+        addSubview(authorLabel)
+        addSubview(timeStampLabel)
+        addSubview(divider)
+        imageView.snp.makeConstraints { (make) in
+            make.centerY.equalToSuperview()
+            make.left.equalToSuperview().inset(18)
+            make.width.height.equalTo(100)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/BookmarkCell.swift]: make these (and other snapkit below) constants as well

File: Cornell Sun/Cornell Sun/Cells/BookmarkCell.swift [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    func setupViews() {
+        self.backgroundColor = .white
+        addSubview(imageView)
+        addSubview(titleLabel)
+        addSubview(authorLabel)
+        addSubview(timeStampLabel)
+        addSubview(divider)
+        imageView.snp.makeConstraints { (make) in
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/BookmarkCell.swift]: remove parentheses on make

File: Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+class PhotoGallerySectionController: ListSectionController {
+    var entry: PostObject!
+
+    override init() {
+        super.init()
+        inset = UIEdgeInsets(top: 0, left: 0, bottom: 15, right: 0)
+    }
+}
+
+extension PhotoGallerySectionController: HeartPressedDelegate, BookmarkPressedDelegate, SharePressedDelegate, PhotoChangedDelegate {
```
Comment (cdsciavolino) [inline Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift]: Can we abstract all of these delegates or reuse code across the other classes that implement these delegates?

File: Cornell Sun/Cornell Sun/Cells/PhotoGalleryCell.swift [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        let fullscreenVC = slideShow.presentFullScreenController(from: getCurrentViewController()!)
+        fullscreenVC.slideshow.activityIndicator = DefaultActivityIndicator(style: .white, color: nil)
+        fullscreenVC.captions = post?.photoGalleryObjects.map({$0.caption})
+        fullscreenVC.captionLabel.text = post?.photoGalleryObjects[slideShow.currentPage].caption
+
+    }
+    func setUpImages() {
+        var kingfisherSource: [KingfisherSource] = []
+        for photoGalleryObject in post!.photoGalleryObjects {
+            let source = KingfisherSource(urlString: photoGalleryObject.fullImageLink)
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/PhotoGalleryCell.swift]: maybe do guard let here

File: Cornell Sun/Cornell Sun/Cells/TitleCell.swift [hunk 15]
```diff
@@ -39,7 +39,7 @@ final class TitleCell: UICollectionViewCell {
         addSubview(titleLabel)
         titleLabel.snp.makeConstraints { (make) in
             make.centerY.equalToSuperview()
-            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 5, left: 17, bottom: 5, right: 5))
+            make.edges.equalToSuperview().inset(UIEdgeInsets(top: 5, left: 17, bottom: 5, right: 17))
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/Cells/TitleCell.swift]: make constants

File: Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+                        cell.bookmarkButton.transform = CGAffineTransform.identity
+        })
+        if didBookmark {
+            RealmManager.instance.save(object: entry)
+        } else {
+            RealmManager.instance.delete(object: entry)
+        }
+    }
+
+    func didPressHeart(_ cell: MenuActionCell) {
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift]: these animations for heart and share could be subclassed for a the button we're reusing (eg. like, bookmark, etc) so making a LikeButton or BookmarkButton

File: Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+            cell.setBookmarkImage(didSelectBookmark: entry.didSave)
+            return cell
+        }
+    }
+
+    override func didUpdate(to object: Any) {
+        entry = object as? PostObject
+    }
+
+    override func didSelectItem(at index: Int) {
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/PhotoGallerySectionController.swift]: remove if it's not being used

File: Cornell Sun/Extensions.swift [hunk 18]
```diff
@@ -7,6 +7,41 @@
...
 
 import UIKit
+import Kingfisher
+
+func captionMaxHeight(width: CGFloat) -> CGFloat {
+     let tmpStr = "A rink attendant collect fish thrown by the Cornell students in a tradition that spans over decades. (Cameron Pollack/ Sun Photography Editor)"
+    return tmpStr.height(withConstrainedWidth: width - 36, font: UIFont(name: "Georgia", size: 13)!)
+}
+
+//func titleMaxHeight(width: CGFloat) -> CGFloat {
```
Comment (mindylou) [inline Cornell Sun/Extensions.swift]: remove if not needed

File: Cornell Sun/Cornell Sun/PostObject.swift [hunk 19]
```diff
@@ -10,80 +10,166 @@ import Foundation
...
+        super.init(realm: realm, schema: schema)
+    }
+
+    required init() {
+        super.init()
     }
 
     func diffIdentifier() -> NSObjectProtocol {
-        return title as NSObjectProtocol
+        var idDiff = -9999999
```
Comment (cdsciavolino) [inline Cornell Sun/Cornell Sun/PostObject.swift]: not quite sure what this -9999999 is for, but you mentioned changing up the realm logic to remove this :)

**Pull request merged**